feat: Add rivet-mini packer, achieving 268 byte executable

realJoshByrnes opened the pull request:
Introduces `rivet-mini`, a custom PE packing tool designed to push binary size limits ensuring functionality on Windows. Key changes:
- Created `rivet-mini` crate for post-processing PE files.
- Implemented header overlapping (PE header moved to offset 4).
- Implemented Optional Header trimming (removing unused data directories).
- Implement aggressive section trimming: removing trailing zeros and leading padding before EntryPoint
- Added logic to shift sections and patch RVAs (EntryPoint, BaseOfCode, etc.).
- Preserved overlay data by merging it into the last section.
- Renamed executables (`msrw.exe`, `msrw-unpacked.exe`, `packer.exe`) to avoid confusion.
- Reduced `msrw.exe` from 448 bytes down to 268 bytes.
- Updated README.md with usage guide and attribution.
- Integrated build step into `rivet-mini` to automatically build `msrw-unpacked`.
- Added `cargo pack` alias for streamlined building and packing.
- Updated CI workflow to verify `cargo pack` and lint the entire workspace.

realJoshByrnes committed:
feat: Add rivet-mini packer, achieving 268 byte executable
Introduces `rivet-mini`, a custom PE packing tool designed to push
binary size limits ensuring functionality on Windows.
Key changes:
- Created `rivet-mini` crate for post-processing PE files.
- Implemented header overlapping (PE header moved to offset 4).
- Implemented Optional Header trimming (removing unused data directories).
- Implement aggressive section trimming: removing trailing zeros and leading padding before EntryPoint
- Added logic to shift sections and patch RVAs (EntryPoint, BaseOfCode, etc.).
- Preserved overlay data by merging it into the last section.
- Renamed executables (`msrw.exe`, `msrw-unpacked.exe`, `packer.exe`) to avoid confusion.
- Reduced `msrw.exe` from 448 bytes down to 268 bytes.
- Updated README.md with usage guide and attribution.
- Integrated build step into `rivet-mini` to automatically build `msrw-unpacked`.
- Added `cargo pack` alias for streamlined building and packing.
- Updated CI workflow to verify `cargo pack` and lint the entire workspace.

realJoshByrnes commented:
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/mcountryman/min-sized-rust-windows/pull/21#pullrequestreview-3556327497)

mcountryman commented on Cargo.toml:
```diff
@@ -1,3 +1,10 @@
+[workspace]
+members = ["rivet-mini"]
+
+[[bin]]
```
Should be able to omit this in favor of `members = [".", "rivet-mini"]`.  The [default-members](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-default-members-field) key should be able to handle the default bin selection.

realJoshByrnes replied:
[root package](https://doc.rust-lang.org/cargo/reference/workspaces.html#root-package):
If the [[workspace] section](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section) is added to a Cargo.toml that already defines a [package], the package is the root package of the workspace.

[default-members](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-default-members-field):
"When unspecified, the [root package](https://doc.rust-lang.org/cargo/reference/workspaces.html#root-package) will be used."

The wording of this has led me to believe we don't need to explicitly list the root package in members, or add a default-members field.

I have made the changes as per your request.

mcountryman replied:
Oh wow TIL, awesome.  My main point there was I think we don't need the `[[bin]]` section though.  I _think_ that still stands

mcountryman replied:
@realJoshByrnes I was referring to lines 4-6 here but, I now see your changes require the binary name to be msrw-unpacked so I suppose it's necessary.

mcountryman commented on rivet-mini/src/main.rs:
```diff
@@ -0,0 +1,532 @@
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+  // Build the project first
+  let status = Command::new("cargo")
```
I'm not a big fan of spawning `cargo` from the packer.  It introduces complexity and limitations that probably can be solved via the env vars mentioned in #18.  I think I'd rather just run two commands.

Also possibly a reference worth looking at: https://github.com/dtolnay/rustversion/blob/master/build/build.rs

realJoshByrnes replied:
One of the reason's I re-wrote it this way was your request to not have an extra build step in #18. Previously it was standalone and input/output were given as arguments to packer.exe.

Having both executables be built (unpacked/packed) makes diagnosing any issues far easier, as the packed version **should** have identical results to the unpacked version - if not, it's a packer issue.

I can attest that having a single build step (using `cargo pack`) is _really_ helpful and much simpler than having two when making changes, although a ps1 file probably could have the same effect.

Ideally `rivet-mini` would evolve into a proper linker, but I'm not really prepared to do that right now tbh

mcountryman replied:
Yeah I was hoping we could do something like `CARGO_RUSTC_WRAPPER="cargo run revo-mini" cargo build` but I'm now remembering Powershell isn't a fan of inline env vars.  

I'm fine with leaving it be for the moment.  I may want to throw together a [Justfile](https://github.com/casey/just) for this and testing in the future.

mcountryman commented on rivet-mini/src/pe.rs:
```diff
@@ -0,0 +1,94 @@
+#[repr(C, packed)]
```
Shhh little robot friend.  Us cool windows kids know what `ImageDosHeader` is.

mcountryman commented on rivet-mini/src/pe.rs:
```diff
@@ -0,0 +1,94 @@
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone)]
+pub struct ImageDosHeader {
+  pub e_magic: u16,      // Magic number
```
nitpick: `///` comments on the properties are nicer IMO.  I heavily use LSP doc-comments and it wont show if the comments are after the field def and not of the `#[doc]` variety.

realJoshByrnes replied:
Valid. Added additional LSP doc comments for the rest of pe.rs

mcountryman commented on rivet-mini/src/main.rs:
```diff
@@ -0,0 +1,532 @@
+    .status()?;
+
+  if !status.success() {
+    return Err(format!("Build failed with exit code: {:?}", status.code()).into());
```
`panic!("...")` does the same thing AFAIK and is less fluff

realJoshByrnes replied:
I initially chose Result because this error is an expected condition (e.g., another project failing to build) rather than a logic bug, and returning Result is generally considered best practice for CLI tools since it provides clean exit codes and works well with ?.

That said, I understand the preference for simplicity here, so I’ll switch to panic! to keep things consistent with the your style.

realJoshByrnes committed:
chore(rivet-mini): apply @copilot review suggestions

Apply all changes suggested by @copilot in the PR review:
- Add module docs and constants (`MIN_WINDOWS_PE_SIZE`, `IMAGE_NUMBEROF_DIRECTORY_ENTRIES`, `IMAGE_NT_OPTIONAL_HDR64_MAGIC`) in `rivet-mini/src/pe.rs`.
- Replace magic padding number with `pe::MIN_WINDOWS_PE_SIZE`.
- Add comprehensive doc comment for `main` describing purpose, strategy, assumptions, and failure cases.
- Validate `input_path`/`output_path` via `fs::symlink_metadata` to avoid symlink/dir surprises.
- Improve "file too small" error to include actual and required sizes.
- Add bounds checks for Optional Header, Data Directories, and Section Table.
- Validate Optional Header magic (PE32+), clamp `number_of_rva_and_sizes` to 16, and warn when clamping.
- Detect last section by file-order (highest raw-data end) before merging overlay.
- Trim trailing zeros and remove unused variables; remove unused start/_end block.
- Validate `file_alignment` (non-zero, power-of-two) and copy packed header fields into aligned locals via `read_unaligned` to avoid UB.

All Copilot review suggestions from the PR review were implemented and `rivet-mini` builds successfully.

realJoshByrnes commented:
I've applied the @copilot suggested changes, after struggling with fmt/clippy errors (and the wrong commit message) we should be good to go now.

realJoshByrnes committed:
chore: apply @mcountryman review suggestions

- Update members and add default-members key to explicitly list the project root in Cargo.toml
- Add LSP doc style comments to `/rivet-mini/src/pe.rs`
- Refactor rivet-mini's `main()` to panic! instead of returning Err

realJoshByrnes commented:
Please ignore the latest @copilot review, I thought I'd see if it was happy after it implemented the changes it suggested (it's not).

There's one unresolved comment in our conversation, so I've allowed edits by maintainers so that you can implement this the way you'd like, as I'm still unsure how you'd like it. I've implemented the other changes you requested.

mcountryman commented:
> I'm still unsure how you'd like it.

Me too lol.  Assuming I'm correct about the `[[bin]]` section in the Cargo manifest I think all is resolved except that.

realJoshByrnes commented:
> Me too lol. Assuming I'm correct about the `[[bin]]` section in the Cargo manifest I think all is resolved except that.

I did a quick search in the comments `[[bin]]` and can't find that is that you're referring to.
I believe I covered all of your reviews.
